### ISSUE TITLE
How do you use numeric user id with next-auth?

### ISSUE THREAD

**Issue by dillanjwilding** [nextauthjs/next-auth#7966]
### Question 💬

I'm relatively new to NextJS and NextAuth but in other software development it's common to have auto-incrementing primary key IDs for tables. Unfortunately, I couldn't get that working with a simple boilerplate app such as T3 stack with NextAuth as it seems the User id (`DefaultUser` in `next-auth/core/types.d.ts` and `AdapterUser` in `next-auth/adapters.d.ts`) TypeScript type is hard-coded to string so we're forced to use UUIDs.

I tried to extend NextAuth's types in a `next-auth.d.ts` file to support numeric IDs but it doesn't seem like TypeScript supports overwriting types and it made a union type instead.

```
import { DefaultSession, DefaultUser } from 'next-auth'

declare module 'next-auth' {
  interface User extends DefaultUser {
    id: number;
  }
  interface Session {
    user?: {
      id: number;
    } & DefaultSession['user'];
  }
}
```

Which seems to work most places except in [...nextauth].ts where it is giving me this error:

> Type 'string | number' is not assignable to type 'number'. Type 'string' is not assignable to type 'number'.ts(2322)

On the `session.user.id = user.id` line in this section of code:

```
export const authOptions: NextAuthOptions = {
  // Include user.id on session
  callbacks: {
    session({ session, user }) {
      if (session.user) {
        session.user.id = user.id
      }
      return session
    }
  },
  adapter: PrismaAdapter(prisma),
  providers: []
}
export default NextAuth(authOptions)
```

The TypeScript error goes away if I delete the `id: string;` line from the `AdapterUser` from `next-auth/adapters.d.ts` which then falls back to `id: number;` because I set that on User.

```
export interface AdapterUser extends User {
    id: string; // <-- I removed this
    email: string;
    emailVerified: Date | null;
}
```

Ultimately after struggling with it for a while I asked on [StackOverflow](https://stackoverflow.com/questions/74147383/how-to-use-numeric-user-id-with-next-auth?noredirect=1#comment135083764_74147383) and the took the response as a solution (i.e. do what I did and then do a type check to filter out string IDs from the string | number type) and that's what I've been using but I'd like to bring this problem to your attention and help working on getting a more ideal long term solution.

### How to reproduce ☕️

It's been a while and I'm pretty sure you'd run into this problem with a NextJS app using NextAuth with a numeric/integer User ID but how I did it was:

1. npm create t3-app@latest
2. Select TypeScript
3. Enable NextAuth, Prisma, Tailwind, tRPC packages
4. Add NEXTAUTH_SECRET to .env
5. npm run build
6. npm run start

The app starts fine but TypeScript still things User IDs are strings even though they are stored as integers and should be type number so add "next-auth.d.ts" to tsconfig.json "include" array and copy this into the `next-auth.d.ts` file:

```
import { type DefaultSession, type DefaultUser } from 'next-auth'

declare module 'next-auth' {
  interface User extends DefaultUser {
    id: number;
  }
  interface Session {
    user?: {
      id: number;
    } & DefaultSession['user'];
  }
}
```

Now the NextAuthOptions have TypeScript errors. It seems since I encountered this last create-t3-app has done an update and reorganized so I'm seeing type any errors but I'm pretty sure after you clean those up that you'll see the underlying problem about string User IDs not supporting numeric/integer User IDs. I'll try to put together an example repository to showcase the problem.

### Contributing 🙌🏽

Yes, I am willing to help answer this question in a PR

**Comment by leafyshark** [comment]
+1

**Comment by dillanjwilding** [comment]
@leafyshark That's what I have in my project. My post here I tried to replicate again from scratch. You should check out my StackOverflow question about this; it won't solve the underlying NextAuth TypeScript problem but it's a workaround that I've been using in the meantime.

**Comment by leafyshark** [comment]
I decided to go with cuid() string in Prisma schema as this was bothering me...

**Comment by abbasyadollahi** [comment]
Just adding a visual of the resulting merged type for `id` becoming `number | string` when using module augmentation.

Here's my `next-auth.d.ts`.

```typescript
declare module "next-auth" {
  interface User {
    id: number;
  }
}
```

![image](https://github.com/nextauthjs/next-auth/assets/16931654/d36ea925-9146-4733-97bf-086a044bd0d8)

Another issue with module augmentation is that it doesn't allow for property erasure. So in the case above, even if I set my `User` interface to only contain an `id: number` property, there's no way to remove the other attributes (`email`, `name`, `image`) that come from NextAuth's `DefaultUser` interface.

Ideally there'd be a way to tell NextAuth what the definition of a `User` looks like and it could be used throughout all the other provider's fields (for example `user` property in `signIn` callback params, return value in `authorize` callback, etc.). Possibly something as simple as NextAuth leaving the `DefaultUser` blank could work?

**Comment by dillanjwilding** [comment]
If we could provide a User type to NextAuth and they provide a DefaultUser that could satisfy our desire for numeric IDs.

**Comment by dillanjwilding** [comment]
I'd argue your example is bad security, not bad database design patterns. Using auto-incrementing IDs is a relatively common practice and, much like other design patterns, developers must take the appropriate steps to secure their code. I think the NextAuth library shouldn't force you to use any particular way but leave that up to the developers themselves. 

As far as if auto-incrementing IDs are used in the industry/wild, not only was I taught throughout University to use them but also, for better or worse, **_every_** enterprise software company I've worked at since used them.

**Comment by Mark2M** [comment]
Correct… I do the same with SQL and apps on closed networks or where there’s no chance a user can see or manipulate the ID. All I was pointing out is that you shouldn’t expose an auto-incremented PK anywhere that a user can simply try another number. 

In the case I mentioned above the Home Office ignored a report about exposing other users data for two years and didn’t do anything about it until Davy Wynder (if memory serves) contacted them and published the fiasco in PC World magazine. 

I should have been more explicit about the context perhaps…

**Comment by dillanjwilding** [comment]
My pushback is because potential security concerns about bad implementation shouldn't have much weight on wanting to have NextAuth support a standard practice in the industry. With many libraries, the security of the entire codebase isn't inherited from the libraries used.

**Comment by Mark2M** [comment]
Mine was a pointless post perhaps… Just a knee jerk reaction as a couple of places I worked were a little “particular” about that sort of thing, and it was a good few years ago…

**Comment by mp3por** [comment]
Soooo is there a normal solution atm ?

**Comment by dillanjwilding** [comment]
There are a couple of workarounds that I've been using but I agree that the library shouldn't make these kind of decisions for developers. I created this issue hoping to discuss possible solutions and bring it to the attention of the library's maintainers but so far no luck. It's unfortunate that this seems to be the library of choice for facilitating authentication for NextJS and T3 apps and maintainers are unresponsive.

**Comment by hussamkhatib** [comment]
Hi @dillanjwilding, 
You can look at how it's done in [cal.com](https://github.com/calcom/cal.com/blob/c5483c81decfe9640145e0ea8177647d394ad321/packages/types/next-auth.d.ts#L13) 
It's pretty similar to what you shared in stackoverflow.
